internal/wasi/wasi.go

```diff
@@ -306,10 +336,125 @@ type SnapshotPreview1 interface {
+	//
+	// * fd - the file decriptor to get the path of the prestat directory
+	// * path - the offset in `ctx.Memory` to write the result path
+	// * pathLen - FdPrestatDirName writes the result path string to `path` offset for the length of `pathLen`.
+	// Note: The semantics of pathLen may vary depending on wasm runtimes. The semantics in wazero follows that of wasmtime.
+	//       Some runtimes, such as wasmer, consider pathLen to be the maximum length instead of the exact length that should be written.
+	//
+	// FdPrestatDirName returns the following errors.
+	// * ErrnoBadf - if `fd` is invalid
+	// * ErrnoFault - if `path` is an invalid offset due to the memory constraint
+	// * ErrnoNametoolong - if `pathLen` is longer than the actual length of the result path
+	// TODO: FdPrestatDirName may have to return ErrnoNotdir if the type of the prestat data of `fd` is not a PrestatDir.
+	//
+	// For example, if fd 3 is a file with path = "test",
+	// and FdPrestatDirName parameters fd = 3, path = 1, and pathLen = 3, we expect `ctx.Memory` to contain:
+	//
+	//                pathLen
+	//              +---------+
+	//              |         |
+	//   []byte{?, 't', 'e', 's', ?, ?, ?}
+	//       path --^
+	//
+	// Note: `PrNameLen` field of the result of FdPrestatGet has the exact length of the actual path.
+	// See FdPrestatGet
+	// Note: ImportFdPrestatDirName shows this signature in the WebAssembly 1.0 (MVP) Text Format.
+	// See https://github.com/WebAssembly/WASI/blob/snapshot-01/phases/snapshot/docs.md#fd_prestat_dir_name
+	FdPrestatDirName(ctx wasm.ModuleContext, fd uint32, resultPrestat uint32) wasi.Errno
+
 	// TODO: wasi.FdPwrite
+
 	// TODO: wasi.FdRead
+	// FdRead is the WASI function to read from a file descriptor.
+	//
+	// * fd - the file decriptor to read from
+	// * iovs - the offset in `ctx.Memory` that contains a series of `IOVec` struct, which indicates where to write the bytes read. See the example below.
+	// * iovsLen - the number of `IOVec`s in `iovs`
+	// * resultSize - the offset in `ctx.Memory` to write the number of bytes read
+	//
+	// FdRead returns the following errors.
+	// * ErrnoBadf - if `fd` is invalid
+	// * ErrnoFault - if `iovs` or `resultSize` contain an invalid offset due to the memory constraint
+	// * ErrnoIo - if an IO related error happens during the operation
+	//
+	// For example, suppose fd 3 is a file with the contents "test",
+	// and FdRead parameters fd = 3, iovs = 1, and iovsLen = 2, resultSize = 24
+	// and two `IOVec`s are located at the offset 1.
+	// Those two `IOVec`s are as follows.
+	//   []IOVec {
+	//     {
+	//       buf: 18    /* type: uint32 */,
+	//       bufLen: 2  /* type: uint32 */,
+	//     },
+	//     {
+	//       buf: 21    /* type: uint32 */,
+	//       bufLen: 2  /* type: uint32 */,
+	//     },
+	//   }
+	// Note: Wazero does not export nor define IOVec struct explicitly. These are examples.
```

Comment:
let's not define things as a struct. stick with the bit layout. Because I think many times we'll have offset, length pairs in this spec even if the name varies. 

Ex. in text
each element of iovs is a memory offset/length pair encoded sequentially as uint32le

I can see the diagram is getting crowded, and applaud the try. Let's try a different layout for this more complicated one..

In the diagram switch from pseudo-struct, s/IOVec/iovs[N/ and re-optimize the space to use what was used for uint32le to buf, bufLen as it was getting crowded. We can still put a reminder to substitute.

```
        // Note: each buf and bufLen are encoded as uint32le
       //
	//                       iovs[0]                    iovs[1]    
	//              +--------------------+   +--------------------+
	//              |      buf     |   |bufLen  |    |   buf      |   |  bufLen  |
	//              +--------+  +--------+   +--------+  +--------+
	//              |        |  |        |   |        |  |        |
	
```

Reply:
after separating the diagram into the input and the output, the diagram becomes as  simple as Environ and Args's one, so I keep using the byte by byte memory diagram now. Other points are fixes as discussed